File: src/conduit/loader/json_loader.py
"""JSON / JSONL destination writer."""

from __future__ import annotations

import json
import logging
from pathlib import Path

import pyarrow as pa

from conduit.models import DestinationConfig

logger = logging.getLogger(__name__)


def load_json(table: pa.Table, dest: DestinationConfig, base_dir: Path) -> None:
    """Write an Arrow table to a JSON or JSONL file."""
    file_path = base_dir / dest.config.get("path", f"{dest.name}.json")
    file_path.parent.mkdir(parents=True, exist_ok=True)

    orient = dest.config.get("orient", "records")
    jsonl = dest.config.get("jsonl", False) or dest.type == "jsonl"

    rows = table.to_pylist()

    with open(file_path, "w", encoding="utf-8") as f:
        if jsonl:
            for row in rows:
                f.write(json.dumps(row, default=str) + "\n")
        elif orient == "records":
            json.dump(rows, f, indent=2, default=str)
        else:
            json.dump(rows, f, indent=2, default=str)

    logger.info(
        "Loaded %d rows to %s destination '%s' (%s)",
        len(table), "JSONL" if jsonl else "JSON", dest.name, file_path,
    )

File: src/ts/loader/json_loader.ts
/**
 * JSON / JSONL destination writer.
 */

import { join, dirname } from "@std/path";
import type { DataTable, DestinationConfig } from "../models.ts";

export async function loadJson(
  table: DataTable,
  dest: DestinationConfig,
  baseDir: string,
): Promise<void> {
  const filePath = join(baseDir, (dest.config.path as string) ?? `${dest.name}.json`);
  Deno.mkdirSync(dirname(filePath), { recursive: true });

  const jsonl = ((dest.config.jsonl as boolean) ?? false) || dest.type === "jsonl";

  if (jsonl) {
    const lines = table.rows.map((row) => JSON.stringify(row)).join("\n") + "\n";
    await Deno.writeTextFile(filePath, lines);
  } else {
    await Deno.writeTextFile(filePath, JSON.stringify(table.rows, null, 2));
  }

  const format = jsonl ? "JSONL" : "JSON";
  console.log(
    `Loaded ${table.rows.length} rows to ${format} destination '${dest.name}' (${filePath})`,
  );
}
